composer.json

```diff
@@ -37,17 +37,19 @@
         "source": "https://github.com/MyIntervals/emogrifier"
     },
     "require": {
-        "php": "~7.0.0 || ~7.1.0 || ~7.2.0 || ~7.3.0",
+        "php": "~7.1.0 || ~7.2.0 || ~7.3.0",
         "ext-dom": "*",
         "ext-libxml": "*",
         "symfony/css-selector": "^2.8 || ^3.0 || ^4.0"
     },
     "require-dev": {
         "friendsofphp/php-cs-fixer": "^2.15.3",
-        "squizlabs/php_codesniffer": "^3.5.0",
-        "slevomat/coding-standard": "^5.0.4",
         "phpmd/phpmd": "^2.7.0",
-        "phpunit/phpunit": "^5.7.27"
+        "phpunit/phpunit": "^7.5",
+        "psalm/plugin-phpunit": "^0.7.0",
+        "slevomat/coding-standard": "^5.0.4",
+        "squizlabs/php_codesniffer": "^3.5.0",
+        "vimeo/psalm": "^3.5"
```

Comment:
As `psalm/plugin-phpunit` depends on `vimeo/psalm`, do we need this?

Reply:
I'd like to have dependencies explicit here and generally not depend on transitive dependencies.

Reply:
If you ever added maglnet/composer-require-checker to your ci checklist, you'd be prompted to explicitly require it.

Reply:
Hmm... What if `psalm/plugin-phpunit` 0.7.1 decides that it requires `vimeo/psalm` 4.0?  I've seen things like this happen.  Our build would suddenly become broken despite us not changing anything.  That's why I suggest that we only require the immediate dependencies, and let them figure out what they need in turn.  It is however possible to use [suggests](https://getcomposer.org/doc/03-cli.md#suggests) with Composer, which may be the solution here.

Reply:
> What if `psalm/plugin-phpunit` 0.7.1 decides that it requires `vimeo/psalm` 4.0?

then it wouldn't be installed if vimeo/psalm was required as "^3.x"

Reply:
> > What if `psalm/plugin-phpunit` 0.7.1 decides that it requires `vimeo/psalm` 4.0?
> 
> then it wouldn't be installed if vimeo/psalm was required as "^3.x"

That's my point.  Composer would 'error' because it could not satisfy the dependencies.

Reply:
> > > What if `psalm/plugin-phpunit` 0.7.1 decides that it requires `vimeo/psalm` 4.0?
> > 
> > 
> > then it wouldn't be installed if vimeo/psalm was required as "^3.x"
> 
> That's my point. Composer would 'error' because it could not satisfy the dependencies. And even if allowed to continue, the build would fall over.

But that might be exactlty what you want...

Reply:
But you could end up with a mysteriously broken build despite having not made any changes.  Or more likely, you've made a change which has triggered the build and you're tearing you hair out trying to figure out just how you've broken it.

Reply:
But I concede 😣

Reply:
> That's my point. Composer would 'error' because it could not satisfy the dependencies. And even if allowed to continue, the build would fall over.

Not sure how it would error? composer's conflict resolution would see the constraint of ^0.7, see that the hypothetical 0.7.1 requested 4.x & 0.7.0 requested 3.5.x, then see that psalm 3.5 was also required, and thus install psalm 3.5.x alongside the phpunit plugin 0.7.0.

not sure I've understood your argument :s

Reply:
> Not sure how it would error? composer's conflict resolution would see the constraint of ^0.7, see that the hypothetical 0.7.1 requested 4.x & 0.7.0 requested 3.5.x, then see that psalm 3.5 was also required, and thus install psalm 3.5.x alongside the phpunit plugin 0.7.0

Good point.  Struggling for a logical argument, next best I can come up with is that suppose `psalm/plugin-phpunit` v0.8.0 no longer requires `vimeo/psalm`.  Then we are left with an orphaned dependency.

I think I know from experience that having this theoretically unnecessary dependency could lead to problems in future.  Precisely what experience that was I'm finding hard to pinpoint/recollect...

Reply:
> Struggling for a logical argument, next best I can come up with is that suppose `psalm/plugin-phpunit` v0.8.0 no longer requires `vimeo/psalm`.

Not likely in this particular case (it being a plugin for psalm that it essentially suppresses a bunch of issues that crop up when using phpunit, i.e. dataProvider methods being flagged as unused because they're not explicitly called etc.)).

> Precisely what experience that was I'm finding hard to pinpoint/recollect...

I do know what you mean here though- every now and then I have to review all my dependencies to ensure they're actually needed.

Reply:
> > Precisely what experience that was I'm finding hard to pinpoint/recollect...
> 
> I do know what you mean here though- every now and then I have to review all my dependencies to ensure they're actually needed.

This is a more general point about [dependecy hell](https://en.wikipedia.org/wiki/Dependency_hell) (which Composer attempts to address), and the fact that I've experienced the ‘hell’, but right now can’t recall details of specific examples.